memtable.cc

```diff
@@ -536,6 +536,13 @@ class partition_snapshot_accounter {
     // partition_snapshot_reader may compose them. In doing so, we move memory to the standard
     // allocation. As long as our size read here is lesser or equal to the size in the memtables, we
     // are safe, and worst case we will allow a bit fewer requests in.
+    //
+    // FIXME: Fragmented buffers have a per-fragment memory overhead. Since mutation fragments are
+    // allocated using the standard allocator, which has a separate preferred_max_contiguous_allocation(),
+    // they could be more fragmented (and therefore use a little more memory) than their source in memtable.
+    // To prevent that, preferred_max_contiguous_allocation() should be kept bigger for the standard
+    // allocator than for the LSA.
+
     void operator()(const range_tombstone& rt) {
         _accounter.update_bytes_read(rt.memory_usage(_schema));
```

Comment:
Can we instead fix `memory_usage()` to include storage overhead as well?

I'd like to avoid linking these two values in this way, it makes changing them in a correct way harder.

Reply:
`memory_usage()` is accounting for storage overhead, that's not the issue here. The problem is that `partition_snapshot_accounter` is trying to estimate the memory usage in memtables, based on the memory usage of the data copied to mutation fragments, which exist in a different allocator. It is impossible to find out the original memory usage at this point in code. I don't see the big picture well enough to propose a fix to that.

Reply:
@tgrabiec can you advise?

Reply:
If this proves to be difficult, we can defer it to a follow-up.

Reply:
@tgrabiec Bump. Please take a look at this.

Reply:
Underaccounting is a preexisting problem. You could get it also when there are multiple MVCC versions of the row. Only the merged version will be subject to accounting, and thus under-account. We have to under-account in this case because (1) those versions may be merged in the memtable by the time we finish flushing and (2) it's better to underaccount than overaccount (which can lead to OOM because we admit more writes than there is space).

In this case, I think it's ok to underaccount, the difference should be relatively insignificant, it is amortized by the fragment size.

Reply:
Sorry, maybe I should have given more context before summoning you here. The problem is: we are possibly *over*accounting, because the accounted version may be bigger due to the per-fragment overhead (`blob_storage`) of `managed_bytes`, which may be bigger here than in the memtable, depending on how `preferred_max_contiguous_allocation` is set.

Context: after this series, the below assert in `memtable.cc` will fail during unit tests, if `standard_allocation_strategy::preferred_max_contiguous_allocation` is set low enough (1052 bytes is small enough to break the tests, if I recall correctly):
```
~flush_memory_accounter() {
    assert(_mt._flushed_memory <= _mt.occupancy().used_space());
}
```

I don't know what to do about that.